functionalization: map copy_() -> to().expand_as()

bdhirsh opened the pull request:
The functionalization pass doesn't work with `copy_()` which is a problem with functorch. Originally we were going to make a functional `copy()` operator to fix this problem, but @zou3519 that we can get (most of) the existing functionality by mapping `self.copy_(src)` to `src.to(self).expand_as(self)`. This makes the codegen a bit uglier, but has the benefit of avoiding a totally unnecessary tensor allocation in functorch.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #66101 Test functionalization pass in python
* **#67878 functionalization: map copy_() -> to().expand_as()**
* #67715 bugfix for conditional functionalization

Differential Revision: [D32280588](https://our.internmc.facebook.com/intern/diff/D32280588)

bdhirsh committed:
functionalization: map copy_() -> to().expand_as()

[ghstack-poisoned]

zou3519 commented on tools/codegen/gen_functionalization_type.py:
```diff
@@ -232,7 +250,7 @@ def emit_inplace_functionalization_body(
           at::AutoDispatchSkipFunctionalize guard;
           // The functionalization pass explicitly doesn't pass out= parameters to the redispatch
-          tmp_output = at::_ops::{functional_op.func.name.unambiguous_name()}::redispatch({', '.join(functional_exprs)});
+          {functional_call_str}
```
Maybe we should have landed your test PR before making this change because I'm about to suggest some more test cases haha

zou3519 commented on tools/codegen/gen_functionalization_type.py:
```diff
@@ -195,7 +195,22 @@ def emit_inplace_functionalization_body(
+    # - mutation removal is only used by functorch, and not by Vulkan or XLA. Functorch already doesn't preserve strides.
+    # - There are actually a few other places where the functionalization pass currently doesn't support strides:
+    #   calls to slice/diagonal_scatter don't currently preserve the strides of their inputs (but maybe we should fix this).
+    if str(f.func.name) == 'copy_':
+        exprs = [keyset] + [a.name for a in unwrapped_args_ctx]
+        functional_call_str = f"""\
+            auto tmp_intermediate = at::_ops::to_other::redispatch({keyset}, src_, self_, non_blocking, false, c10::nullopt);
+            tmp_output = at::_ops::expand_as::redispatch({keyset}, tmp_intermediate, self_);"""
```
There are some interesting cases that we should test.
- A copy_ case where the dtype and shape are the same (this is already present in the functionalization PR)
- A copy_ case where the dtype is different
- A copy_ case where the shape is different
- Because `expand_as` returns a view and an expanded tensor, we should check that further mutation (in the original code) of the result still produces the correct output.

So, for example:
```
def f(x):
  tmp = torch.zeros(2, 2)
  tmp_slice = tmp.diagonal()
  r1 = tmp_slice.copy_(x)
  r2 = r1.add_(x)
  return x, r1, r2

cases = [
  torch.tensor(1.),
  torch.tensor(1, dtype=torch.long),
  torch.randn(2),
]
```

bdhirsh replied:
oof thanks for pointing out https://github.com/pytorch/pytorch/pull/66101#discussion_r743713239, and that also pretty clearly points out the extra coverage for `copy_()` that you're suggesting (in some of your other examples, I **would** expect `to` to show up in the LoggingTensor trace). I'll add those in before landing the stack.

zou3519 commented on tools/codegen/gen_functionalization_type.py:
```diff
@@ -195,7 +195,22 @@ def emit_inplace_functionalization_body(
+    # We could add one, but that would be sub-optimal for functorch: copy() would need to allocate a fresh tensor.
+    # This may seem like a large hack for one optimization, but copy_() is one of the most common inplace operators.
+    # Instead, we can replace `self.copy_(src)` with `src.to(self).expand_as(self)`.
+    # This maintains the exact same semantics, EXCEPT that we don't preserve the strides from `self`.
```
sidenote: I think we already knew this, but in general functionalization doesn't preserve the "same semantics" as the original function. This is probably OK but we might need to figure out how to better explain that to whoever uses it. Here's an example:
```
def foo(x, y):
  return x.add_(y)

x = torch.randn(1)
y = torch.randn(3)

# Produces an error in normal PyTorch
foo(x, y)

# Inside functionalization, x.add_(y) becomes x + y, and no error is produced!
# Also: I guess this means that torch/xla has slightly different semantics from eager-mode PyTorch as well?
functionalize(foo)(x, y)
```

bdhirsh replied:
Yep, good point, and these differences probably mostly match up with pytorch/xla's differences from cpu eager mode (although pytorch/xla has a few more differences due to bugs, that will get fixed by the pass).

zou3519 reviewed (approved):
I think the test cases I suggested will be fine, but those are good to have

bdhirsh committed:
Update on "functionalization: map copy_() -> to().expand_as()"

The functionalization pass doesn't work with `copy_()` which is a problem with functorch. Originally we were going to make a functional `copy()` operator to fix this problem, but @zou3519 that we can get (most of) the existing functionality by mapping `self.copy_(src)` to `src.to(self).expand_as(self)`. This makes the codegen a bit uglier, but has the benefit of avoiding a totally unnecessary tensor allocation in functorch.




[ghstack-poisoned]

bdhirsh committed:
Update on "functionalization: map copy_() -> to().expand_as()"

The functionalization pass doesn't work with `copy_()` which is a problem with functorch. Originally we were going to make a functional `copy()` operator to fix this problem, but @zou3519 that we can get (most of) the existing functionality by mapping `self.copy_(src)` to `src.to(self).expand_as(self)`. This makes the codegen a bit uglier, but has the benefit of avoiding a totally unnecessary tensor allocation in functorch.




[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32280588).
